### PR TITLE
fix: resolve issue where marshal fails with cross api dependency

### LINKED ISSUE #151: Unable to configure config management with Clusters registered with Anthos
#### Environment details

  - OS type and version: Windows 10
  - Python version: 3.8.5
  - pip version: 22.0.2 
  - google-cloud-gke-hub 1.4.2

#### Steps to reproduce

  1. Trying to create configure config management feature in one of the cluster registered with Anthos with the below code

#### Code example

```
from google.cloud import gkehub_v1
from google.oauth2 import service_account
from google.cloud.gkehub_v1.types import MembershipFeatureSpec
from google.cloud.gkehub_v1.types import feature
from google.cloud.gkehub_v1 import configmanagement_v1

projectId = 'xxxxxx'
location = 'global'


def create_feature(client, parent, name):
    git_config = configmanagement_v1.GitConfig(
        sync_repo="https://github.com/GoogleCloudPlatform/anthos-config-management-samples",
        sync_branch="main",
        policy_dir="multi-environments-kustomize/config-source/overlays/dev",
        secret_type=None
    )
    config_sync = configmanagement_v1.ConfigSync(
        git=git_config,
        source_format='unstructured'
    )
    configmanagement = configmanagement_v1.MembershipSpec(
        config_sync=config_sync,
        version='1.11.2'
    )
    member_spec = MembershipFeatureSpec(
        configmanagement=configmanagement
    )
    feature_spec = feature.Feature(
        name=name,
        membership_specs={
            name: member_spec

        }
    )
    # Initialize request argument(s)
    request = gkehub_v1.CreateFeatureRequest(
        parent=parent,
        feature_id='configmanagement',
        resource=feature_spec
    )

    # Make the request
    operation = client.create_feature(request=request)

    print("Waiting for operation to complete...")

    response = operation.result()
    print(response)

    # Handle the response
    print(response.pb)

cluster_name = 'xxxxxxxxxx'
credentials = service_account.Credentials.from_service_account_file('credentials.json')
client = gkehub_v1.GkeHubClient(credentials=credentials)
parent = 'projects/%s/locations/%s' % (projectId, location)
membership_name = '%s/memberships/%s' % (parent, cluster_name)
create_feature(client, parent, membership_name)

```

#### Stack trace
```
Traceback (most recent call last):
  File "list_config.py", line 60, in <module>
    create_feature(client, parent, membership_name)
  File "list_config.py", line 26, in create_feature
    member_spec = MembershipFeatureSpec(
  File "\pythonProjectAnthos\venv\lib\site-packages\proto\message.py", line 585, in __init__
    super().__setattr__("_pb", self._meta.pb(**params))
TypeError: Parameter to MergeFrom() must be instance of same class: expected google.cloud.gkehub.configmanagement.v1.MembershipSpec got MembershipSpec.
```

### LINKED ISSUE #349: Marshal fails with TypeError when using cross api message
See test code here https://github.com/googleapis/proto-plus-python/pull/348/commits/e123ad8813823c477d1ff1494a1c6c1ba85bff89, with corresponding failure [here](https://github.com/googleapis/proto-plus-python/actions/runs/4426034682/jobs/7761831001).

See the full stack trace below
```
________________________ test_module_package_cross_api _________________________

    def test_module_package_cross_api():
        sys.modules[__name__].__protobuf__ = proto.module(package="spam.eggs.v1")
        try:
    
            class Baz(proto.Message):
                foo = proto.RepeatedField(proto.INT64, number=1)
    
    
            marshal = proto.Marshal(name="spam.eggs.v1")
    
            assert Baz.meta.package == "spam.eggs.v1"
            assert Baz.pb() in marshal._rules
    
            sys.modules[__name__].__protobuf__ = proto.module(package="ham.pancakes.v1")
    
            class AnotherMessage(proto.Message):
                qux = proto.Field(proto.MESSAGE, number=1, message=Baz)
    
            marshal = proto.Marshal(name="ham.pancakes.v1")
    
            assert AnotherMessage.meta.package == "ham.pancakes.v1"
            assert AnotherMessage.pb() in marshal._rules
            # Confirm that Baz.pb() is no longer present in marshal._rules
            assert Baz.pb() not in marshal._rules
    
            # Test using multiple packages together
>           msg = AnotherMessage(qux=Baz())

tests/test_modules.py:65: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError('Unknown field for AnotherMessage: _pb') raised in repr()] AnotherMessage object at 0x7f112bbc07d0>
mapping = {'qux': }, ignore_unknown_fields = False, kwargs = {'qux': }
params = {'qux': }
marshal = <proto.marshal.marshal.Marshal object at 0x7f112bbc0910>, key = 'qux'
value = , pb_value = 

    def __init__(
        self,
        mapping=None,
        *,
        ignore_unknown_fields=False,
        **kwargs,
    ):
        # We accept several things for `mapping`:
        #   * An instance of this class.
        #   * An instance of the underlying protobuf descriptor class.
        #   * A dict
        #   * Nothing (keyword arguments only).
        if mapping is None:
            if not kwargs:
                # Special fast path for empty construction.
                super().__setattr__("_pb", self._meta.pb())
                return
    
            mapping = kwargs
        elif isinstance(mapping, self._meta.pb):
            # Make a copy of the mapping.
            # This is a constructor for a new object, so users will assume
            # that it will not have side effects on the arguments being
            # passed in.
            #
            # The `wrap` method on the metaclass is the public API for taking
            # ownership of the passed in protobuf object.
            mapping = copy.deepcopy(mapping)
            if kwargs:
                mapping.MergeFrom(self._meta.pb(**kwargs))
    
            super().__setattr__("_pb", mapping)
            return
        elif isinstance(mapping, type(self)):
            # Just use the above logic on mapping's underlying pb.
            self.__init__(mapping=mapping._pb, **kwargs)
            return
        elif isinstance(mapping, collections.abc.Mapping):
            # Can't have side effects on mapping.
            mapping = copy.copy(mapping)
            # kwargs entries take priority for duplicate keys.
            mapping.update(kwargs)
        else:
            # Sanity check: Did we get something not a map? Error if so.
            raise TypeError(
                "Invalid constructor input for %s: %r"
                % (
                    self.__class__.__name__,
                    mapping,
                )
            )
    
        params = {}
        # Update the mapping to address any values that need to be
        # coerced.
        marshal = self._meta.marshal
        for key, value in mapping.items():
            (key, pb_type) = self._get_pb_type_from_key(key)
            if pb_type is None:
                if ignore_unknown_fields:
                    continue
    
                raise ValueError(
                    "Unknown field for {}: {}".format(self.__class__.__name__, key)
                )
    
            try:
                pb_value = marshal.to_proto(pb_type, value)
            except ValueError:
                # Underscores may be appended to field names
                # that collide with python or proto-plus keywords.
                # In case a key only exists with a `_` suffix, coerce the key
                # to include the `_` suffix. It's not possible to
                # natively define the same field with a trailing underscore in protobuf.
                # See related issue
                # https://github.com/googleapis/python-api-core/issues/2[27](https://github.com/googleapis/proto-plus-python/actions/runs/4426034682/jobs/7761831001#step:6:28)
                if isinstance(value, dict):
                    if _upb:
                        # In UPB, pb_type is MessageMeta which doesn't expose attrs like it used to in Python/CPP.
                        keys_to_update = [
                            item
                            for item in value
                            if item not in pb_type.DESCRIPTOR.fields_by_name
                            and f"{item}_" in pb_type.DESCRIPTOR.fields_by_name
                        ]
                    else:
                        keys_to_update = [
                            item
                            for item in value
                            if not hasattr(pb_type, item)
                            and hasattr(pb_type, f"{item}_")
                        ]
                    for item in keys_to_update:
                        value[f"{item}_"] = value.pop(item)
    
                pb_value = marshal.to_proto(pb_type, value)
    
            if pb_value is not None:
                params[key] = pb_value
    
        # Create the internal protocol buffer.
>       super().__setattr__("_pb", self._meta.pb(**params))
E       TypeError: Parameter to MergeFrom() must be instance of same class: expected spam.eggs.v1.Baz got Baz.

proto/message.py:[60](https://github.com/googleapis/proto-plus-python/actions/runs/4426034682/jobs/7761831001#step:6:61)4: TypeError
```

### PR DESCRIPTION
**Pull request by parthea**
Fixes #349 
Fixes https://github.com/googleapis/python-gke-hub/issues/151
Googlers see go/proto-plus-cross-api-issue

### CODE REVIEW

**Comment by parthea** [comment]
Waiting for approval on go/proto-plus-cross-api-issue

**Review by atulep — APPROVED** [review]
LGTM

File: tests/test_modules.py
Comment (vchudnov-g) [inline tests/test_modules.py:65]: So the error would show up instantiation time in line 64, right? Could you add a brief comment to that effect, with either the error we would get or at least a link to https://github.com/googleapis/python-gke-hub/issues/151?

(without the context it's a bit confusing)

**Review by vchudnov-g — APPROVED** [review]

File: tests/test_modules.py
Comment (parthea) [reply tests/test_modules.py:65]: Correct, the stack trace is in the bug that is linked: https://github.com/googleapis/proto-plus-python/issues/349. I'll add a link to the bug above line 64.

Comment (parthea) [reply tests/test_modules.py:66]: Done.

**Pull request merged**
